Add totality checking

wilbowma opened the pull request:
This PR replaces #103.

pwang347 committed:
Initial iteration

pwang347 committed:
Minor

pwang347 committed:
Update

pwang347 committed:
newlines

pwang347 committed:
nit

pwang347 committed:
Integration, better warnings

pwang347 committed:
Minor

pwang347 committed:
Infer types, format-id

pwang347 committed:
Re-write, bindings okay with subst, TODO merge wildcards

pwang347 committed:
Minor remove print

pwang347 committed:
Is this it... do I see the light?

pwang347 committed:
Fix temp numbers

pwang347 committed:
wip

pwang347 committed:
fix most of the remaining issues

pwang347 committed:
Comments

pwang347 committed:
Matcher complete, type inference bug with nested generics

pwang347 committed:
Possibly last commit

wilbowma committed:
Merge branch 'pwang-totality-check' of https://github.com/pwang347/cur into pr-103-totality-checking

wilbowma committed:
Style fixes and documentation

wilbowma committed:
Fix naming conventions to avoid exposing Turnstile

Also, some whitespace fixes

wilbowma committed:
Style fixes and documentation

wilbowma committed:
Simplfying pattern tree implementation

pwang347 committed:
test fix

pwang347 committed:
oops lol

pwang347 committed:
Update error message

pwang347 committed:
And now, I've come full circle

wilbowma committed:
Style fixes and documentation

wilbowma committed:
Replicate Paul's suggested fixes to test locally

Seems to work

wilbowma committed:
Reviewing code

wilbowma committed:
TODOs

wilbowma committed:
Remove some unnecessary reparsing

wilbowma committed:
Merge branch 'pr-103-totality-checking' of https://github.com/pwang347/cur into pr-103-totality-checking

wilbowma commented:
Opened this to easily see conflicts/Travis in one place.

Looks like we're getting false negatives from Travis. It passes, but I see the following in the logs and when I run `Poly-pairs.rkt` myself.

```
raco test ../../../cur/cur-test/cur/tests/ntac/software-foundations/Poly-pairs.rkt

Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
Failed to determine type of #<syntax temp105>
ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
 also, no #%top syntax transformer is bound
  in: temp105 #<continuation-mark-set> (#<syntax temp105>))
raco test: "../../../cur/cur-test/cur/tests/ntac/software-foundations/Poly-pairs.rkt"
raco test: @(test-responsible '(wilbowma stchang))
32 tests passed

wilbowma commented:
cc @pwang347

wilbowma committed:
Reviewing pattern-tree

pwang347 commented:
> Opened this to easily see conflicts/Travis in one place.
> 
> Looks like we're getting false negatives from Travis. It passes, but I see the following in the logs and when I run `Poly-pairs.rkt` myself.
> 
> ```
> raco test ../../../cur/cur-test/cur/tests/ntac/software-foundations/Poly-pairs.rkt
> 
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> Failed to determine type of #<syntax temp105>
> ERROR: #(struct:exn:fail:syntax temp105: unbound identifier;
>  also, no #%top syntax transformer is bound
>   in: temp105 #<continuation-mark-set> (#<syntax temp105>))
> raco test: "../../../cur/cur-test/cur/tests/ntac/software-foundations/Poly-pairs.rkt"
> raco test: @(test-responsible '(wilbowma stchang))
> 32 tests passed
> ```

I made a brief comment about this in #103: 

> Currently there is a bit of log spewing from get-typeof in stdlib\pattern-tree.rkt for certain files. This is because it doesn't know how to handle implicitly defined constructors, especially trickier cases like :: which supports infinite nesting (this is the main TODO that I'm aware of)

I had it left as is, but maybe it might be preferred to silence it and add a TODO instead. (The error itself is technically not wrong as it failed to produce a type, but there are a few specific cases where it's unhandled at this point)

wilbowma commented:
Missed that, thanks!

pwang347 commented:
Also if I remember correctly, the behaviour of untyped variables in totality checking is just an automatic pass, so totality checking should be skipped for any and all variables facing such an issue (potentially a partial totality check, whew that sounded weird)

pwang347 commented:
https://github.com/wilbowma/cur/pull/105/commits/4b0faf18bf121c0b10d55665d885128a813db30a#diff-286c6976832f9a510328591747a86955R125

I think I also wanted to call subst-tmp on the terms here whoops

wilbowma committed:
Fix typo in stlc test

wilbowma commented:
> 
> 
> Also if I remember correctly, the behaviour of untyped variables in totality checking is just an automatic pass, so totality checking should be skipped for any and all variables facing such an issue (potentially a partial totality check, whew that sounded weird)

Didn't really understand this comment. Do you mean if we have a variable missing a type annotation, totality checking will automatically succeed?

pwang347 commented:
So say we match on two variables, e.g. `a, b`
Where both are intended to be Nat, but for whatever reason b wasn't able to be labelled.
The following match cases should pass:
```
[z z]
[(s x) z]
```

But the following should not:
```
[z z]
[z (s x)]
```

This is usually not a problem for positional arguments that are passed in by the user (in this case there is no ambiguity on the type), but it becomes an issue when we generate temporaries while flattening the patterns, since the type of the temporaries are inferred from the nested match expressions.

wilbowma commented:
Hm. That's a strange behavior and makes me worry about type soundness. Will have to think on it.

pwang347 commented:
See updated comment above! Yeah, can potentially lead to some false negatives. I think I was trying to play it safe with backwards compat in this case.

wilbowma commented:
Normally a good call, but in a dependently typed language, soundness trumps backwards compatibility.

pwang347 commented:
Makes sense. For reference, the behaviour is created here:
https://github.com/wilbowma/cur/pull/103/files#diff-df1a362f622a4c434e1c207377ada588R23
(`get-constructors-metadata` returns false if no type information is inferred, which implies there's nothing to check)

wilbowma commented:
Thanks!

wilbowma committed:
Simplify some pattern-tree code

wilbowma committed:
Style and documentation in pattern-tree

wilbowma committed:
Add failure when checking totality finds no constructors to check against

wilbowma committed:
Reviewing pattern-tree

wilbowma commented:
Looks like the first failing definition is this one:
```

(define/rec/match split_ : [X : Type] [Y : Type] (list (prod X Y))
  -> (prod (list X) (list Y))
  [nil => (pair (nil* X) (nil* Y))]
  [(:: (pair m n) xs)
   => (pair (:: m (fst (split_ X Y xs)))
            (:: n (snd (split_ X Y xs))))]
  #:type-aliases ([pair = pair* 2]))

```
It ought to be total, so this is a false positive resulting from lack of type information that we ought to be able to recover.

wilbowma commented:
Okay, I think I'm declaring this out of scope for the current PR. The root cause of the problem seems to be poor handling or implicits, but the implementation of implicits is flawed and at the root of the problem.

wilbowma committed:
Reviewing pattern-tree

wilbowma committed:
Reviewing pattern-tree

wilbowma committed:
Work-around issue with totality checker and implicits

wilbowma committed:
Update data to define-datatype in Poly-pairs

wilbowma committed:
Note about weird test in Poly-pairs

pwang347 commented:
Hey I'm still out doing a big move out from my apartment, but I'll try to address all your TODO comments when I'm back.

Edit: probably won't be back until close to midnight

wilbowma commented:
No worries, I'm still reviewing and thinking about the code. Not sure how many will remain. I'm removed at least one as I understood the code better.

wilbowma committed:
Update data to define-datatype in Tactics-destruct

wilbowma committed:
Work-around issue with totality checker and implicits

wilbowma committed:
Whitespace

wilbowma committed:
Disable some tests that I think should be invalid

wilbowma committed:
Can we use fresh instead of generate-temporary?

I think fresh gives us something syntactically the same but with a different scope, which is nice for printing and testing

wilbowma committed:
Reviewing more code

wilbowma committed:
Simplify pattern processing hash

Replaced pattern processing logic with a custom-hash and updater, to avoid
manually traversing and search for a key in the hash.
This should be faster, and is more direct about what's going on.

wilbowma committed:
Replace mutable C-hash with immutable

wilbowma committed:
Reorder some things to match documentation

wilbowma committed:
Simplify pattern variable check

wilbowma committed:
Style and documentation in pattern-tree

wilbowma committed:
Documentation for pattern tree

wilbowma committed:
Simplify is-constructor?

wilbowma committed:
reviewing

pwang347 commented on cur-lib/cur/stdlib/sugar.rkt:
```diff
@@ -184,53 +187,86 @@
+               ([ty-in-aliases (~datum =) ty-out-aliases
+                               (~optional
+                                ty-out-implicit-counts:exact-nonnegative-integer
+                                ;; TODO PR103: Why is this a syntax-object, while others
```
I used this because I wanted a default argument, but as was the case with me not knowing that `~and` was a thing, I'm sure things can be cleaned up quite a bit here

pwang347 commented on cur-lib/cur/stdlib/sugar.rkt:
```diff
@@ -184,53 +187,86 @@
       (~datum :)
-      (~optional decreasing-arg:exact-nonnegative-integer #:defaults ([decreasing-arg #'0]))
+      (~optional decreasing-arg:exact-nonnegative-integer #:defaults
+                 ;; TODO PR103: Why is this a syntax-object, while others are not?
```
See above

pwang347 commented on cur-lib/cur/stdlib/sugar.rkt:
```diff
@@ -184,53 +187,86 @@
+          #`(define/rec/match^ name : decls-x ... ty-to-match ...
+              decls-y ... -> ty_out clauses ...)
+          'top-level null)
+         ;; TODO PR103: What happens if all fail? Surely the function wouldn't be
```
You're right, it would be better to have an explicit failure maybe. I think the reasoning was that if another error occurred, then it would have raised an exception already. Reaching the end here would be an error in implementation of the termination checker. (see comment on line 233)

pwang347 commented on cur-lib/cur/stdlib/sugar.rkt:
```diff
@@ -266,6 +302,18 @@
                         (begin0 t1-recur-ok?
                                 (unless t1-recur-ok?
                                   (if (syntax-property t2 'last-decreasing-check)
+                                      ;; TODO PR103: This error is confusing. Seems to
```
Agreed

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,699 @@
+  (require racket/dict)
+  (define (create-pattern-tree-decl-helper match-vars pattern-matrix bodies env)
+    (let* (; create a hash to store results for the leading pattern column
+           ; TODO PR103: Is this hash purely an optimization? Looks like an
```
The hash is to determine uniqueness, I think the idea is that we want to merge keys at this step.
For example, if two match cases start with the same match expression then they should be merged into a single node.

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,699 @@
+                                   (lambda (t)
+                                     (and t
+                                          (syntax-property
+                                           ; TODO PR103: Can we use fresh here instead?
```
I don't think I looked into this yet, might be worth a try

wilbowma commented:
I pushed two major changes. One replaced the C-hash implementation with a custom immutable hash. This seemed to work. The second replaced the constructor metadata implementation. This kind of works, but there's a few differences in interface that break in a few places that I haven't finished debugging. Will continue later this week.

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,717 @@
+  ;; hash or merges the remaining patterns to form a new pattern sub-matrix.
+  (define (process-pattern match-var head-pattern remaining-patterns body idx
+                           C-hash env)
+    ; TODO PR103: Not sure if this comment is relevant to the custom hash.
```
Yeah, so in the case of efficiency it's only minor pruning at best. The hash acts more like a set of 2-tuples, where it's convenient to retrieve the value by a key.

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,717 @@
+                  ;; TODO PR103: This is probably wrong now since I changed the
+                  ;; behavior to fail when matching on a type without
+                  ;; constructors.
+                  ;; Doesn't seem to be so that much be some other default case.
```
Did you mean "must" here?

pwang347 replied:
I'll also have to recollect a bit more on why this default was needed

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,717 @@
+    (let ([pattern-as-list (syntax->list pattern)])
+      (if (false? pattern-as-list)
+          ;; TODO PR103: Odd that in this case we return a list, which suggest
+          ;; the input was a valid pattern, but it wasn't?
```
I think it was a valid pattern, except the form of it isn't list-like. I've used this style a few times and I can see how it's a bit confusing with all the `false?` and `list?` being thrown around.

pwang347 commented on cur-lib/cur/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,717 @@
+          env
+          (let* ([constructor-id (first head-pattern-as-list)]
+                 [metadata (get-constructor constructor-id match-var #:env env)]
+                 ;; TODO PR103: All these ands could be short-circuited. If
```
Agreed

pwang347 commented on cur-test/cur/tests/stdlib/pattern-tree.rkt:
```diff
@@ -0,0 +1,740 @@
+
+  ;; WITH AND WITHOUT TYPE CONTEXT
+  ; effectively, z = _ = n in this scenario
+  ; TODO PR103: Don't think this should pass as n is unbound.
```
A possible argument for keeping this is that unbound identifiers do not necessarily need to be caught at this phase of re-write, given that the main purpose of the pattern-tree compiler is to restructure the input syntax

wilbowma replied:
I agree in principle, but it looks like much of the implementation is dependent on Cur specific APIs, and Cur is pretty sensitive about unbound identifiers.

wilbowma committed:
Reviewing

wilbowma committed:
Reviewing

wilbowma committed:
Updating CIC inductive reflection interface

wilbowma committed:
Rewriting pattern-tree reflection code

wilbowma commented:
Okay I'm feeling pretty good about my understanding of the code now. The only major thing I want to change is the interface to inductive constructors. I made some changes, but plan to make some more which should take some O(n) checks down to O(1).

pwang347 commented:
> Okay I'm feeling pretty good about my understanding of the code now. The only major thing I want to change is the interface to inductive constructors. I made some changes, but plan to make some more which should take some O(n) checks down to O(1).

Sounds great, thanks William!

wilbowma committed:
Removed some TODOs

wilbowma committed:
Try using fresh instead of generate-temporaries

wilbowma committed:
Fix some fragile tests

wilbowma committed:
Fix some fragile tests

wilbowma committed:
Fixing tests that rely on names

wilbowma committed:
Fix generated names and whitespace in pattern-tree tests

wilbowma committed:
Remove get-typeof which was just a wrapper

wilbowma committed:
Fix

wilbowma committed:
Trying to delete code

wilbowma committed:
TODOs

wilbowma committed:
Replace function with library function

wilbowma committed:
TODOs

wilbowma committed:
Slightly improve totality error message

wilbowma committed:
Remove dead code

wilbowma committed:
TOdone

wilbowma committed:
Improve error handling in define/rec/match for totality failure

wilbowma committed:
TODOs

wilbowma committed:
TODOs

wilbowma committed:
TODOs

wilbowma committed:
Fix whitespace

wilbowma committed:
Improve error message

wilbowma committed:
Update expected error messages in tests

wilbowma commented:
Closed by 9b3dc826660a21fdd14690a18cad0eaf5e301086

stchang commented:
Is something like this supposed to work?

```racket
(define/rec/match test : (X : Type) X -> X
  [x => x])
```

It's currently rejected by `total?` because `X` is not inductive.

wilbowma commented:
I would say "no", since we can't know anything about that type, such as what it's constructors are. What if at run-time, you provide a function?

While this specific case isn't a problem since it's just the identity, the semantics should be similar to eliminators, and we wouldn't ever be able to eliminate that.

stchang commented:
So Coq would reject this? (cant get Coq running at the moment)

dmelcer9 commented:
Is there an equivalent to def/rec/match in coq?

wilbowma commented:
`def/rec/match` is meant to mimic `Fixpoint` in Coq, IIRC (which requires that you then use `match` in the body)

dmelcer9 commented:
But in 

`match foo, bar with ...`

Both foo and bar need to be inductive vars, while the first line of a def/rec/match is serving that role here.

wilbowma commented:
Right. And the definition using `def/rec/match` is being rejected because the equivalent of `bar` isn't inductive.
